File: src/chat/MyClient.java
package chat;

import sockets.Client;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.file.Files;
import java.nio.file.Paths;

public class MyClient {
    private static Client client;
    private static boolean stop = false;

    public static void main(String[] args) {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String ip;
        String username;

        try {
            System.out.println("Enter the ip address of the Server");
            ip = br.readLine();

            client = new Client(ip, 42069);

            System.out.println("Enter your Username: ");
            username = br.readLine();
            client.sendMessage(username);

        } catch (IOException e) {
            e.printStackTrace();
        }

        showInstructions();

        Thread reader = new Thread(new ClientReader());
        reader.start();
        Thread writer = new Thread(new ClientWriter());
        writer.start();
    }

    private static void stopThread() throws IOException {
        client.close();
        stop = true;
    }

    private static void showInstructions() {
        System.out.println("*****************INSTRUCTIONS*****************");
        System.out.println("Write \"/leave\" to leave the chat");
        System.out.println("Write \"/file\" to send a file");
        System.out.println("You have to type in the directory of the file");
        System.out.println("**********************************************");
    }


    private static class ClientReader implements Runnable {

        @Override
        public void run() {
            String msg;
            while (true) {
                try {
                    if (stop) {
                        return;
                    }

                    msg = client.receiveMessage();

                    switch (msg) {
                        case "/file":
                            receiveFile();
                            break;
                        default:
                            System.out.println(msg);
                            break;
                    }

                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }

        private void receiveFile() throws IOException {
            String filename;

            Files.createDirectories(Paths.get("download"));

            if (Files.notExists(Paths.get("download"))) {
                filename = client.receiveFile();
            } else {
                filename = client.receiveFile("download/");
            }

            System.out.println("The file \"" + filename + "\" was downloaded");
        }
    }

    private static class ClientWriter implements Runnable {

        @Override
        public void run() {
            System.out.println("You can now write messages to the server");

            BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
            String msg;

            while (true) {
                try {
                    msg = br.readLine();

                    switch (msg) {
                        case "/leave":
                            System.out.println("Stopping...");
                            client.sendMessage(msg);
                            stopThread();   //kill other thread
                            return;
                        case "/file":
                            sendFile();
                            break;
                        case "/instruction":
                            showInstructions();
                            break;
                        default:
                            client.sendMessage(msg);
                            break;
                    }

                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }

        private void sendFile() throws IOException {
            BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
            System.out.println("Enter the file path:");
            String filename = br.readLine();

            client.sendMessage("/file");

            if (Files.notExists(Paths.get(filename))) {
                System.err.println("File not found");
                return;
            }

            client.sendFile(filename);
        }
    }
}

File: src/c#/chat/MyClient.cs
﻿using System;
using System.IO;
using System.Net.Sockets;
using System.Threading;
using SN_Chat.sockets;

namespace SN_Chat.chat {
    internal static class MyClient {
        private static Client _client;

        private static void Main() {
            Console.WriteLine("Enter the ip address of the Server:");
            var ipAddress = Console.ReadLine();

            _client = new Client(ipAddress, 42069);

            Console.WriteLine("Enter your Username:");
            var username = Console.ReadLine();

            _client.SendMessage(username);

            ShowInstructions();

            var reader = new Thread(Reader);
            var writer = new Thread(Writer);
            reader.Start();
            writer.Start();
            reader.Join();
            writer.Join();
        }

        private static void ShowInstructions() {
            Console.WriteLine("*****************INSTRUCTIONS*****************");
            Console.WriteLine("Write \"/leave\" to leave the chat");
            Console.WriteLine("Write \"/file\" to send a file");
            Console.WriteLine("You have to type in the directory of the file");
            Console.WriteLine("**********************************************");
        }

        private static void StopThread() {
            _client.Stop();
        }

        private static void Reader() {
            while (true) {
                string msg;
                
                try {
                    msg = _client.ReceiveMessage();
                }
                catch (SocketException) {
                    Console.Error.WriteLine("Socket was closed");
                    return;
                }


                switch (msg) {
                    case "/file":
                        ReceiveFile();
                        break;
                    default:
                        Console.WriteLine(msg);
                        break;
                }
            }
        }

        private static void Writer() {
            Console.WriteLine("You can now write messages to the server");

            while (true) {
                var msg = Console.ReadLine();

                switch (msg) {
                    case "/leave":
                        Console.WriteLine("Stopping...");
                        StopThread();
                        return;
                    case "/file":
                        SendFile();
                        break;
                    case "/instruction":
                        ShowInstructions();
                        break;
                    default:
                        _client.SendMessage(msg);
                        break;
                }
            }
        }

        private static void SendFile() {
            Console.WriteLine("Enter the file path:");
            var filename = Console.ReadLine();

            if (!File.Exists(filename)) {
                Console.WriteLine("File not found");
                return;
            }

            _client.SendMessage("/file");

            _client.SendFile(filename);
        }

        private static void ReceiveFile() {
            Directory.CreateDirectory("download");

            var filename = _client.ReceiveFile(Directory.Exists("download") ? "download/" : "");

            Console.WriteLine("The file \"{0}\" was downloaded", filename);
        }
    }
}